### PR TITLE
CSHARP-4957: Fix issue with new array expression.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs [hunk 1]
```diff
@@ -24,14 +27,29 @@ internal static class NewArrayInitExpressionToAggregationExpressionTranslator
...
+
+                // make sure all items are serialized using the same serializer
+                if (!itemTranslation.Serializer.Equals(itemSerializer))
+                {
+                    throw new ExpressionNotSupportedException(expression, because: "all items in the array must be serialized using the same serializer");
+                }
             }
+
             var ast = AstExpression.ComputedArray(items);
-            var serializer = context.KnownSerializersRegistry.GetSerializer(expression);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs:33]: KnownSerializer was returning the wrong serializer here.

I think we need to reevaluate whether the known serializer strategy is working.

In this particular case we should just determine the required result serializer here without help from the known serializer registry.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs [hunk 2]
```diff
@@ -24,14 +27,29 @@ internal static class NewArrayInitExpressionToAggregationExpressionTranslator
...
         {
             var items = new List<AstExpression>();
+            IBsonSerializer itemSerializer = null;
             foreach (var itemExpression in expression.Expressions)
             {
                 var itemTranslation = ExpressionToAggregationExpressionTranslator.Translate(context, itemExpression);
                 items.Add(itemTranslation.Ast);
+                itemSerializer ??= itemTranslation.Serializer;
+
+                // make sure all items are serialized using the same serializer
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs:37]: This goes slightly beyond fixing the reported bug, but it's in the same logic path.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs [hunk 3]
```diff
@@ -24,14 +27,29 @@ internal static class NewArrayInitExpressionToAggregationExpressionTranslator
...
+                if (!itemTranslation.Serializer.Equals(itemSerializer))
+                {
+                    throw new ExpressionNotSupportedException(expression, because: "all items in the array must be serialized using the same serializer");
+                }
```
Comment (JamesKovacs) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs:41]: Items must use compatible serializers but not necessarily the same serializer. For example I would expect the following to work:

```
var queryable = collection.AsQueryable()
    .Select(x => new Animal[] { x.Cat, x.Dog });
```

**Review by JamesKovacs — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs (hunk 3 above)
Comment (JamesKovacs) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewArrayInitExpressionToAggregationExpressionTranslator.cs:41]: After investigating further and discussing in Slack, we are going to leave this as a known limitation that we don't support arrays of polymorphic types in projections.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**